build: Bump g++ minimum supported version to 12

maflcko opened the pull request:
All supported operating systems that previously came with at least g++-11, also come with at least g++-12, so bumping the minimum should be fine.

For reference:

* https://packages.ubuntu.com/jammy/g++-12
* https://packages.ubuntu.com/noble/g++ (g++-13)
* https://packages.debian.org/bookworm/g++ (g++-12)
* FreeBSD Ports ship a recent GCC
* RHEL-based 8, and 9 ship with g++-14 via appstream (`dnf install gcc-toolset-14` -> `/opt/rh/gcc-toolset-14/`)
* RHEL-based 10 ships with g++ (14 by default)
* OpenSuse Leap and Tumbleweed ship with g++ 15 https://software.opensuse.org/package/gcc15-c++

Obviously, downloading pre-compiled releases or compiling previous release branches is unaffected by this change.

Commit:
ci: Enable experimental kernel stuff in G++-12 task (previous releases)

Base the task on --preset=dev-mode to ensure maximal coverage and add
the following:

   bitcoin-chainstate (experimental) ... ON
   libbitcoinkernel (experimental) ..... ON
   kernel-test (experimental) .......... ON

Also, shorten the name, for a less cluttered web view.

Commit:
build: Bump g++ minimum supported version to 12

hebasto reviewed (commented):
Such a change might also require updating the build guides. For instance, the following commands on Ubuntu 22.04 would use GCC 11:
```
sudo apt-get install build-essential ...
cmake --build build
```

Commit:
util: [refactor] Remove unused create_directories workaround

This was added in commit 1f46b6e46e1454b91ff7ceb31853bc440952f8eb, but
is no longer needed after g++-12 is the minimum required.

maflcko commented:
> Such a change might also require updating the build guides. For instance, the following commands on Ubuntu 22.04 would use GCC 11:

I am happy to add a general note that the latest stable or LTS is recommended and that earlier releases may need the user to install the required minimum dependencies.

Commit:
doc: Recommend latest Debian stable or Ubuntu LTS

sedited reviewed (approved):
ACK fabf8e51083120883162399293c620f2afc0c5e2

hebasto reviewed (approved):
ACK fabf8e51083120883162399293c620f2afc0c5e2.

janb84 reviewed (commented):
lgtm ACK fabf8e51083120883162399293c620f2afc0c5e2

With the note added I can't think of a reason not to bump the g++ version to version 12.

Commit:
test: Remove unused fs::create_directories test

The test was added in commit ddb75c2e87a60ed24065bdf0c3bfabf4e058cef1.
After the create_directories wrapper removal, the test is redundant with
the unit test in the upstream stdlib. Also, there is a Bitcoin Core
functional test that covers this behavior in
test/functional/feature_dirsymlinks.py

So remove this unit test.

Finally, I could not find a real system that still ships a buggy stdlib
(v11.2) in their package manager. A stand-alone test is also available
in compiler-explorer under https://godbolt.org/z/aeMKraYrT.

Commit:
test: Remove gccbug_90348 test case

The test case no longer detects this specific issue for GCC versions
12.1+, as explained in the
https://gcc.gnu.org/bugzilla/show_bug.cgi?id=90348 thread and in this
compiler-explorer playground:
https://godbolt.org/z/Y48osrjM8

So remove the test case and update the -fstack-reuse=none cmake
docstring with the underlying affected GCC versions, and the bug URL.

Commit:
doc: Add GCC-12 min release notes

sedited reviewed (approved):
Re-ACK fa9f29a4a79944f6ffbb58eab0ac41e243fbeb97

hebasto reviewed (approved):
ACK fa9f29a4a79944f6ffbb58eab0ac41e243fbeb97.
